presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java

```diff
@@ -54,7 +55,7 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider type, Plan
         requireNonNull(variableAllocator, "variableAllocator is null");
         requireNonNull(idAllocator, "idAllocator is null");
 
-        if (preferMergeJoin(session)) {
+        if (isGroupedExecutionEnabled(session) && preferMergeJoin(session)) {
```

Comment:
My concern about this is that just because grouped execution is enabled, it doesn't mean the query will run with grouped execution.  I'm not sure how else to check because the complicated logic for figuring it out happens in the PlanFragmenter, but I worry that there will be some cases with invalid plans because run with grouped execution.

Also, I didn't quite understand why bucketed execution is a requirement, but I think grouped execution can still run multiple buckets at a time.  is that okay?

Reply:
Yes, I have a [follow-up PR](https://github.com/prestodb/presto/pull/17890) to address this concern, but I'm still having trouble finding out good test cases (they're pretty rare cases) which is why I didn't attach that commit here.

The goal of this PR is still to guard most of the cases which is better than having no guard at all, the following up PR would guarantee we won't have loose end.


As for why grouped execution is a requirement
1. [Hard blocker]Pairing requirement: when we do merge join, we need to know the pairing information (eg: left table bucket1 has to join with right table bucket1), grouped execution would do that for us (`LifeSpan` in `Split`), without it we need to add additional pairing info (See key word Identifier in this [doc](https://docs.google.com/document/d/1qQGqVU8E1vOSRH_cEEoOfXFyBY-XsV9UWf86iPkuGU0/edit?usp=sharing))
2. Scheduling convenience, it's a bonus feature where it gives us control on when the merge join can start by controlling concurrent_life_span and guarantee performance consistency for each run (See key word scheduling in this [doc](https://docs.google.com/document/d/1qQGqVU8E1vOSRH_cEEoOfXFyBY-XsV9UWf86iPkuGU0/edit?usp=sharing))

Reply:
that makes sense. It would be nice to have a way to prevent creating merge join plans that captures 100% of non-grouped execution cases.  I don't have any suggestions right now though, so this is a good start.